sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala

```diff
@@ -250,7 +276,13 @@ object PartitioningUtils {
       val rawColumnValue = columnSpec.drop(equalSignIndex + 1)
       assert(rawColumnValue.nonEmpty, s"Empty partition column value in '$columnSpec'")
 
-      val literal = inferPartitionColumnValue(rawColumnValue, typeInference, timeZone)
+      val literal = if (userSpecifiedDataTypes.contains(columnName)) {
+        // SPARK-26188: if user provides corresponding column schema, process the column as String
+        //              type and cast it as user specified data type later.
+        inferPartitionColumnValue(rawColumnValue, false, timeZone)
```

Comment:
can't we add the cast here?

Reply:
See my reasons above.

Reply:
can't we make it returning `Option[(String, Literal)]`? If not, what about `Literal(Cast(...).eval())`?

Reply:
Thanks. I will use the Cast one.